Add LG ThinQ Integration

LG-ThinQ-Integration opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add LG ThinQ Integration

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34169

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

JangwonLeeLG committed:
Add manifest.json

YunseonPark-LGE committed:
add switch entity

JangwonLeeLG committed:
Add tests

YunseonPark-LGE committed:
fix function's name

NahyunLee-LGE committed:
adjust the changes after running scipt

LG-ThinQ-Integration committed:
Merge branch 'home-assistant:dev' into pr1_1

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    Platform.SWITCH,
+]
+
+SERVICE_GET_DEVICE_PROFILE_SCHEMA = vol.Schema(
+    {vol.Required(ATTR_DEVICE_ID): cv.string}
+)
+SERVICE_GET_DEVICE_STATUS_SCHEMA = vol.Schema({vol.Required(ATTR_DEVICE_ID): cv.string})
+SERVICE_POST_DEVICE_STATUS_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_DEVICE_ID): cv.string,
+        vol.Required(SERVICE_ATTR_VALUE): vol.Any(cv.string, dict),
+    }
+)
+
```
```suggestion
```

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    # Set up all platforms for this device/entry.
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Set up custom services.
+    async_setup_hass_services(hass, entry)
+
```
```suggestion
```

LG-ThinQ-Integration replied:
Removed

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    entry.runtime_data.device_map.clear()
+
+    # Get a device list from the server.
+    response: ThinQApiResponse = await thinq.async_get_device_list()
```
Why is the type cast here? Isn't the upstream library typed?

LG-ThinQ-Integration replied:
Yes. ThinQApiResponse type is defined in PyPI library just like the other integrations do.

frenck commented on homeassistant/components/lgthinq/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "loggers": ["pycountry", "thinqconnect"],
+  "requirements": [    
+    "pycountry==23.12.11",
+    "thinqconnect==0.9.2"
```
Same comment as before: https://github.com/home-assistant/core/pull/123341/files#r1711423028

I could not find the public / open-source repository providing this package. pypi doesn't have any public repository linked, and couldn't find it with a search on GitHub either.

LG-ThinQ-Integration replied:
Please refer to the link below.
- https://pypi.org/project/thinqconnect/

joostlek replied:
But where is it on GitHub?

LG-ThinQ-Integration replied:
@joostlek 
We are planning to open the repository soon.

LG-ThinQ-Integration replied:
We are planning to release it in October.

LG-ThinQ-Integration replied:
Let me share this with someone in charge.
In that case, I think  we can reschedule the release.

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+        old_unique_id = next(iter(old_entry.identifiers))[1]
+        if old_unique_id not in new_device_unique_ids:
+            device_registry.async_remove_device(old_entry.id)
+            _LOGGER.warning("Remove device_registry: device_id=%s", old_entry.id)
```
```suggestion
            _LOGGER.debug("Remove device_registry: device_id=%s", old_entry.id)
```

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+        if device is None:
+            return async_create_service_response(device_id, None, None)
+
+        value = call.data.get(SERVICE_ATTR_VALUE)
+        result = await device.async_post_device_status(value)
+        return async_create_service_response(device_id, device, result)
+
+    hass.services.async_register(
+        domain=DOMAIN,
+        service="reload_device_list",
+        service_func=async_handle_reload_device_list,
+    )
+    hass.services.async_register(
+        domain=DOMAIN,
+        service="refresh_device_status",
+        service_func=async_handle_refresh_device_status,
+    )
+    hass.services.async_register(
+        domain=DOMAIN,
+        service="get_device_profile",
+        service_func=async_handle_get_device_profile,
+        schema=SERVICE_GET_DEVICE_PROFILE_SCHEMA,
+        supports_response=SupportsResponse.ONLY,
+    )
+    hass.services.async_register(
+        domain=DOMAIN,
+        service="get_device_status",
+        service_func=async_handle_get_device_status,
+        schema=SERVICE_GET_DEVICE_STATUS_SCHEMA,
+        supports_response=SupportsResponse.ONLY,
+    )
+    hass.services.async_register(
+        domain=DOMAIN,
+        service="post_device_status",
+        service_func=async_handle_post_device_status,
+        schema=SERVICE_POST_DEVICE_STATUS_SCHEMA,
+        supports_response=SupportsResponse.ONLY,
+    )
+
+
```
```suggestion
```

LG-ThinQ-Integration replied:
Removed

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    )
+
+
+@callback
+def async_create_service_response(
+    device_id: str | None,
+    device: LGDevice | None,
+    result: dict[str, Any] | str | None,
+) -> ServiceResponse:
+    """Create a service response from the result of service call."""
+    if result is None:
+        if device_id is None:
+            result = "error: No device_id specified."
+        elif device is None:
+            result = "error: Device not found."
+        else:
+            result = "error: Operation failed."
+
+    device_info: dict[str, JsonValueType] = {}
+    if device is not None:
+        device_info["id"] = device.id
+        device_info["sub_id"] = device.sub_id
+        device_info["type"] = device.type
+        device_info["name"] = device.name
+        device_info["model"] = device.model
+
+    return {
+        ATTR_DEVICE_ID: device_id,
+        SERVICE_ATTR_DEVICE_INFO: device_info,
+        SERVICE_ATTR_RESULT: result,
+    }
+
+
```
```suggestion
```

LG-ThinQ-Integration replied:
Removed.

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+
+    async def _validate_and_create_entry(self) -> ConfigFlowResult:
+        """Create an entry for the flow."""
+        connect_client_id: str = f"{CLIENT_PREFIX}-{uuid.uuid4()!s}"
```
String type is already inferred

```suggestion
        connect_client_id = f"{CLIENT_PREFIX}-{uuid.uuid4()!s}"
```

LG-ThinQ-Integration replied:
String type is removed.

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+        result = await api.async_get_device_list()
+        _LOGGER.debug("validate_and_create_entry: %s", result)
+
+        if result.status >= 400:
```
Shouldn't the library raise an error/exception in these cases? Seems a bit odd. As I cannot find the sources of the library, I'm not sure why this is the way it is here.

LG-ThinQ-Integration replied:
This is just to handle REST API error cases.
Other integrations are doing it in the same way.

LG-ThinQ-Integration replied:
Removed

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+        _LOGGER.debug("validate_and_create_entry: %s", result)
+
+        if result.status >= 400:
+            # Support translation for TRANSLATION_ERROR_CODE, key is error_code.
```
This makes our translation keys coupled with an upstream library. We should be doing that. Please decouple it.

LG-ThinQ-Integration replied:
Is it really necessary? Other integrations are using it when they handle REST API error cases.

LG-ThinQ-Integration replied:
Removed

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+                data_schema=vol.Schema(
+                    {
+                        vol.Required(CONF_ACCESS_TOKEN): cv.string,
+                        vol.Optional(CONF_NAME, default=THINQ_DEFAULT_NAME): cv.string,
```
Please do not ask for an config entry name in the configuration flow.

```suggestion
```

LG-ThinQ-Integration replied:
I refer to other integrations' config flow.
Do you have any changes in rules of config flow?

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+            return self.async_abort(reason=reason_str)
+
+        # If verification is success, create entry.
+        data = {
+            CONF_ACCESS_TOKEN: self._access_token,
+            CONF_CONNECT_CLIENT_ID: connect_client_id,
+            CONF_COUNTRY: self._country_code,
+        }
+        return self.async_create_entry(title=self._entry_name, data=data)
```
There is no need to assign this to a variable first.

```suggestion
        return self.async_create_entry(
            title=self._entry_name,
            data={
                CONF_ACCESS_TOKEN: self._access_token,
                CONF_CONNECT_CLIENT_ID: connect_client_id,
                CONF_COUNTRY: self._country_code,
            }
        )
```

frenck commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        # Get the PAT(Personal Access Token) and validate it.
+        if user_input is None or CONF_ACCESS_TOKEN not in user_input:
```
CONF_ACCESS_TOKEN is always in the user input... the schema required it.

```suggestion
        if user_input is None:
```

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    # Initialize runtime data.
+    entry.runtime_data = ThinqData()
+
+    thinq = ThinQ(
+        client_session=async_get_clientsession(hass),
+        country_code=entry.data.get(CONF_COUNTRY, DEFAULT_COUNTRY),
+        client_id=client_id,
+        access_token=access_token,
+    )
```
Are there any possible side-effect from this?

LG-ThinQ-Integration replied:
I guess not. We are handling error cases when the integration communicate with the PyPI library.
If you want to discuss further about this, we better do this later with the last PR.

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+        client_id=client_id,
+        access_token=access_token,
+    )
+    device_registry = dr.async_get(hass)
```
Not fully understanding why the device registry is retrieved in this location. Probably better to move this into the places it is actually used (since you pass along `hass` anyways, you can do it locally).

LG-ThinQ-Integration replied:
We thought we better call it once than twice.

LG-ThinQ-Integration replied:
Moved into the each method.

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+            translation_key=response.error_code,
+        )
+
+    device_list = response.body
```
body? What is that?

LG-ThinQ-Integration replied:
It is json result object (dictonary) for device list.

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+        )
+
+    device_list = response.body
+    if not device_list or not isinstance(device_list, Collection):
```
Why is an `ininstance` check needed here?

LG-ThinQ-Integration replied:
We used it to handle REST API error case becase of the ThinQ Cloud.

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    entry: ThinqConfigEntry,
+) -> None:
+    """Register devices to the device registry."""
+    if lg_device_list:
```
This check shouldn't be needed, according to the method signature above, this is always a list.

```suggestion
```

frenck commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+) -> None:
+    """Clean up device registry."""
+    new_device_unique_ids: list[str] = [device.unique_id for device in lg_devices]
+    existing_entries: list[dr.DeviceEntry] = dr.async_entries_for_config_entry(
```
This is already typed upstream?

```suggestion
    existing_entries = dr.async_entries_for_config_entry(
```

LG-ThinQ-Integration replied:
Removed.

frenck commented on homeassistant/components/lgthinq/const.py:
```diff
@@ -0,0 +1,126 @@
+THINQ_PAT_URL: Final[str] = "https://connect-pat.lgthinq.com"
+
+# Config Flow
+CLIENT_PREFIX: Final[str] = "home-assistant"
+CONF_CONNECT_CLIENT_ID: Final[str] = "connect_client_id"
+DEFAULT_COUNTRY: Final[str] = "US"
```
There are tons more of these cases to be found across this PR. Please review them all and remove the ones that can be inferred.

LG-ThinQ-Integration replied:
We thought Home Assistant was very strict at type definition. So we refered to some recent merged PRs to make our code correct like below. 
- https://github.com/home-assistant/core/pull/123056/files

LG-ThinQ-Integration replied:
Removed.

frenck reviewed (dismissed):
Some (second time) initial review comments to get the review process going. It is not a full review, as many aspects can be removed or cleaned up across the PR.

../Frenck

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+class ThinqData:
+    """A class that holds runtime data."""
+
+    device_map: dict[str, LGDevice] = field(default_factory=dict)
```
@frenck 
ThinqData class has now 1 item, because the other item related to mqtt event is removed for thefirst PR.
So, we need to keep it.

LG-ThinQ-Integration replied:
Removed.

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Accept the suggested change about format.

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Accept suggested change for log removal

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Delete homeassistant/components/lgthinq/services.yaml

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/services.yaml:
@frenck 
The service file is removed

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/switch.py

Accpet suggested change for log removal

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/strings.json

Accept suggested change for service removal

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/manifest.json

Accept suggested change for spaces removal

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Delete homeassistant/components/lgthinq/icons.json

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/icons.json:
removed

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/thinq.py:
The thinqconnect PyPI communicates with ThinQ Cloud.
And this class is handing the data conversion or managing just like smartthings integration's 'api' or 'smartapp' class.
This class doens't communicate with ThinQ Cloud directly.
So I think we don't need to put this class into PyPI library.

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration committed:
Update __init__.py

Remove unnecessary check code

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,331 @@
+    # Validate entry data.
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
+    if not isinstance(client_id, str):
+        raise ConfigEntryError(f"Invalid client id: {client_id}")
```
It is unnecessary. Removed.

LG-ThinQ-Integration commented:
@frenck 
I resolved all the reviews, please check those again.

LG-ThinQ-Integration committed:
Merge branch 'dev' into pr1_1

JangwonLeeLG committed:
Modification to pass ruff-format

LG-ThinQ-Integration committed:
Modification for mypy issues

LG-ThinQ-Integration committed:
Remove service registry and related code

LG-ThinQ-Integration committed:
Update strings.json

Modification to pass the prettier issues

LG-ThinQ-Integration committed:
Update manifest.json

Modification to pass the prettier issues

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Remove the unnecessary log.

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Remove unnecessary log.

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Remove unnecessary code.

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/property.py:
'thinqconnect' PyPI library communicates with ThinQ Cloud.
This just handle property data handling or conversion like 'smartapp' or 'api' in 'SmartThings' integration.

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Modifications for the review and related autocheck

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/config_flow.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/config_flow.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Modifications for reviews and autocheck

LG-ThinQ-Integration committed:
Modifications for the reviews and autocheck

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+            )
+
+        self._access_token = str(user_input.get(CONF_ACCESS_TOKEN, ""))
+        self._entry_name = str(user_input.get(CONF_NAME, ""))
```
Other integrations are getting entry_name from the config flow.
Is there any change in your guidance?

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+        self._entry_name = str(user_input.get(CONF_NAME, ""))
+
+        # Check if PAT is already configured.
+        await self.async_set_unique_id(self._access_token)
```
Yes. it can.

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+
+        return await self.async_step_region()
+
+    async def async_step_region(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
```
We refere to other integrations. (ex. smartthings, tuya, etc)
We just want to make a step as simple as possible.
Is it mandatory to put things together in a single step?

LG-ThinQ-Integration replied:
We put this step into 'user' step.

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+                },
+            )
+
+        self._access_token = str(user_input.get(CONF_ACCESS_TOKEN, ""))
```
The access token is created in ThinQ Cloud web page.
And it is used all the time on communicating with cloud through REST API.
So it has to be the part of the schema as required.

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/const.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,146 @@
+    def __init__(self) -> None:
+        """Initialize flow."""
+        super().__init__()
+        self._access_token: str = ""
+        self._country_code: str = DEFAULT_COUNTRY
+        self._entry_name: str = ""
```
If we remove those, it is detected by pylint W0201.
As I mentioned already in other reviews, entry_name is used in other integrations' config flow.

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/const.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/const.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/device.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/device.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/const.py:
```diff
@@ -0,0 +1,126 @@
+SERVICE_ATTR_RESULT: Final[str] = "result"
+SERVICE_ATTR_VALUE: Final[str] = "value"
+
+# Types for profile and property.
+type Profile = dict[str, Any]
+type ProfileMap = dict[str, Profile]
+type PropertyMap = dict[str, ProfileMap]
```
We hope we can use them.

LG-ThinQ-Integration replied:
Removed

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/const.py:
```diff
@@ -0,0 +1,126 @@
+
+
+# Error Code
+class ErrorCode(StrEnum):
```
Please let us move it before the last PR.

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/entity_helpers.py:
```diff
@@ -0,0 +1,3147 @@
+from thinqconnect.const import DeviceType
+from thinqconnect.thinq_api import ThinQApiResponse
+
+from homeassistant.components.binary_sensor import BinarySensorEntityDescription
+from homeassistant.components.button import ButtonEntityDescription
+from homeassistant.components.climate import ClimateEntityDescription
+from homeassistant.components.event import EventEntityDescription
+from homeassistant.components.fan import FanEntityDescription
+from homeassistant.components.humidifier import (
+    HumidifierDeviceClass,
+    HumidifierEntityDescription,
+)
+from homeassistant.components.number import (
+    NumberDeviceClass,
+    NumberEntityDescription,
+    NumberMode,
+)
+from homeassistant.components.select import SelectEntityDescription
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntityDescription,
+    SensorStateClass,
+)
+from homeassistant.components.switch import SwitchEntityDescription
+from homeassistant.components.text import TextEntityDescription
+from homeassistant.components.time import TimeEntityDescription
+from homeassistant.components.vacuum import StateVacuumEntityDescription
+from homeassistant.components.water_heater import WaterHeaterEntityEntityDescription
```
As you mentioned, we already limited a real functionality to a single platform.
But because we support almost 30 types of devices and 10 entities. (Home Appliances).
Limiting every code into a single platform per PR is very complicated based on our code.
And it will slow us down so much.
This file is mapping entities to devices.
Could you review this file in the last PR?

LG-ThinQ-Integration commented:
@frenck 
I resolved all your reviews and autochecks.

LG-ThinQ-Integration committed:
Merge branch 'dev' into pr1_1

LG-ThinQ-Integration committed:
Merge branch 'dev' into pr1_1

LG-ThinQ-Integration committed:
Remove type definition after Final

LG-ThinQ-Integration committed:
Update const.py

Do not use Final for DOMAIN

LG-ThinQ-Integration committed:
Merge branch 'home-assistant:dev' into pr1_1

LG-ThinQ-Integration committed:
Merge branch 'home-assistant:dev' into pr1_1

JangwonLeeLG committed:
Refactoring for reviews
- remove thinq.py
- remove type definition
- remove entry name in config flow
- put config flow steps into a single step

JangwonLeeLG committed:
Update tests
- remove region

LG-ThinQ-Integration commented:
@frenck 
Thank you for your reviews. 
I resolved all your reviews except a few things which we need to keep.
Please check those again.

joostlek commented on homeassistant/components/lgthinq/device.py:
```diff
@@ -0,0 +1,439 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class LGDevice:
```
Okay so what exactly is the role of the LGDevice in the integration?

LG-ThinQ-Integration replied:
1. We are planning to release it in October.
2. A LGDevice represents a single LG appliance device which is different from typical IoT devices.
We need to handle devices' data, notification and status through LG Device instance.
It's because of ThinQ Cloud structures which is also communicating with mobile apps.
It also works as a coordinator for a device.

joostlek replied:
> A LGDevice represents a single LG appliance device which is different from typical IoT devices

I don't really see a difference in how it's different from typical IoT devices and would love to get to know more about this so I can get in the same flow of decision making. I think using the common practices we have in Home Assistant is more maintainable in the long run.

Is it possible to maybe chat on Discord with one of you?

joostlek commented on homeassistant/components/lgthinq/device.py:
```diff
@@ -0,0 +1,439 @@
+        # Get the api instance.
+        self._api: ConnectBaseDevice = api.get_sub_device(self._sub_id) or api
+
+        # Create property map form the given api instance.
+        self._property_map: dict[str, dict[str, dict[str, Any]]] = (
+            self._retrieve_profiles(self.api.profiles)
+        )
```
What;s this?

LG-ThinQ-Integration replied:
Move into the PyPI library.

joostlek commented on homeassistant/components/lgthinq/entity_helpers.py:
```diff
@@ -0,0 +1,312 @@
+}
+
+
+class ThinQEntity(CoordinatorEntity, Generic[ThinQEntityDescriptionT]):
```
Why is it generic? You can just overwrite the base entity_description type

LG-ThinQ-Integration replied:
Removed.

joostlek commented on homeassistant/components/lgthinq/property.py:
I am not sure what I think of this file. It contains a lot of service specific information like on how to post values and how to parse them. I think that that is part of the device specific communication, and thus belongs in the library

LG-ThinQ-Integration replied:
Moved into PyPI library.

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,69 @@
+) -> None:
+    """Set up an entry for switch platform."""
+
+    @callback
+    def async_add_devices(devices: Collection[LGDevice]) -> None:
+        """Add switch entities."""
+        async_add_entities(ThinQSwitchEntity.create_entities(devices))
+
+    async_add_devices(entry.runtime_data.device_map.values())
+
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, THINQ_DEVICE_ADDED, async_add_devices)
+    )
```
I think the integration could be way more maintainable if it was using the common architecture of an integration.

`async_setup_entry` in `__init__.py` -> Everything needed to setup the integration
`async_setup_entry` in the platforms -> Everything needed to add entities and to add them on runtime
The entity classes -> These should be plain classes which only transform the state they represent from the state of the integration to the state they represent.

A coordinator is a common way to hold state of a (part of the) integration. Those should exist in `coordinator.py`. Adding new entities or devices is not a responsibility of the entities (like the `ThinQSwitchEntity.create_entities` here) but a responsibility of the `async_setup_entry`s at the platform.

LG-ThinQ-Integration replied:
Refactored.
Some code is moved into the PyPI library.

JangwonLeeLG committed:
Merge branch 'dev' of https://github.com/LG-ThinQ-Integration/core into pr1_1

JangwonLeeLG committed:
Refactoring for reviews
- move property.py into PyPI library
- replace error_code handling with try/catch
- remove http response handling
- remove generic
- remove unnecessary class or map instance
- refactor adding entities logic

JangwonLeeLG committed:
Refactoring
- remove unused code
- change import path

JangwonLeeLG committed:
Update tests

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,151 @@
+    # Validate entry data.
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
+    if not isinstance(client_id, str):
+        client_id = f"{CLIENT_PREFIX}-{uuid.uuid4()!s}"
```
1. ThinqData is removed.
2. The unnecessary conditional code is removed.

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,151 @@
+        client_id = f"{CLIENT_PREFIX}-{uuid.uuid4()!s}"
+    access_token = entry.data.get(CONF_ACCESS_TOKEN)
+    if not isinstance(access_token, str):
+        raise ConfigEntryAuthFailed(f"Invalid PAT: {access_token}")
```
Removed.

LG-ThinQ-Integration commented:
@joostlek 
I resolved everything I can. 
Even though I can't make some reviews resolved, I did what you expected to do.
Please check those again.

joostlek commented:
I left a response on one of them and I am still awaiting this one https://github.com/home-assistant/core/pull/123860#discussion_r1722967090

LG-ThinQ-Integration commented:
@joostlek 
I left comments about timeline on the original review.

joostlek commented:
https://github.com/home-assistant/core/pull/123860#discussion_r1732558234 and https://github.com/home-assistant/core/pull/123860#discussion_r1732548139

LG-ThinQ-Integration commented:
@joostlek 
Unfortunately, we cant' use Discord for security reasons.
Let me explain more about it.
The basic reason is that thinq cloud which thinqconnect library communicates with works not per entity but per device.
That is because thinqconnect library is designed for several environments including web/android applications etc.
If we provide simple functions as IoT devices, it could be same.
But we want to provider as many functions as our devices provide.
So we need to control things per device.
To be specific, a entity can have dependency to others, and we need to set or get status together per device.
ex) if we change 'course mode' in Washer, default 'Rinse' profile is changed and 'Spin Speed' profile is also changed.
To communicate with thinq cloud, we need this concept.
If not, we have to seperate all the communications into entity in the integration.
I think it is time-consuming for us.

joostlek commented:
> Unfortunately, we cant' use Discord for security reasons

Understood.

So let me try to summarize so I understand correctly (please confirm or correct me so I know I understood):
- You fetch data per device, which data is exposed with multiple entities
- That data fetching is centralized

LG-ThinQ-Integration commented:
You are correct.

LG-ThinQ-Integration commented:
And about the timeline, I left a comment.
I understand and respect your policy.
We will reschedule it before PR in September.

joostlek commented:
Awesome! I will write something down about how the architecture ideally would work which would make it very maintainable and nice. In the meantime I have a question that would help the explanation:

- How do we know if there is a new device or when a device has been removed? Do we poll every so often?

joostlek commented:
Thanks for your understanding on the library :)

LG-ThinQ-Integration commented:
We don't use polling. 
We have mqtt protocol which is not included in the first PR.
(Actually it is removed for frenck's reviews)
It will let us know the status update, device removal/adding, events and etc.

joostlek commented:
Check. Another question popped up, are the entities that devices expose static or can they register on runtime (for a single device)?

joostlek commented:
So a common architecture we see is the following:

- `__init__.py` is responsible for the following
   - Checking if the integration is able to setup (this can be checked by fetching data)
   - Creating coordinators for every device we want to add (an example can be found at https://github.com/home-assistant/core/blob/3336bdb4026f87e4d9676acba682c85033f2ac14/homeassistant/components/yolink/__init__.py#L149-L164 I am not completely happy with how this looks, but please use it as example for understanding the concept)
   - Forward the setup to the platforms
- `coordinator.py` contains an extended coordinator.
    - I think this is similar to what your LGdevice now is, but if we extend a DataUpdateCoordinator, we already have quite some functionality covered that the LGDevice is adding now. Making the code cleaner and more recognizable.
    - We overwrite `_async_update_data` to return the state of that device (example: https://github.com/home-assistant/core/blob/e68901235b411189e8309f17329264872afbbae7/homeassistant/components/analytics_insights/coordinator.py#L38-L61)
- `entity.py` contains a base entity
    - This base entity will be extending CoordinatorEntity. It will take care of a couple of basic things like the device_info, `_attr_has_entity_name = True` and possible availability checks. (Example: https://github.com/home-assistant/core/blob/c80718628ef81abd3ee069db53bed8008deb4aa3/homeassistant/components/airgradient/entity.py#L10-L20)
 - The platform files like `switch.py`
     - `async_setup_entry` is responsible for deciding which entities should be created and added. The entities should be passed to `async_add_entities` and should not be stored anywhere else.
     - So in general, every platform just gets a list of a few devices, and should decide "does this device need a switch?". (Example: https://github.com/home-assistant/core/blob/ab9cbf64da39fab4583fd1c0becc159342ab0b9a/homeassistant/components/airgradient/sensor.py#L250-L264 You can see that we only add certain entities based on their model)
 
 
 
 Since the automatic device adding has been removed, I would propose that we keep the logic required for that out of the picture for now. That logic should live in both `__init__.py` and `async_setup_entry` of the platform files.

And please don't see this as a negative thing, the integration is quite near what I am proposing here, but it just requires some small refactors. If something is unclear or if you have questions, please feel free to ask. I have a bunch of examples to explain concepts and design patterns that we can use to make this integration neat and extendable.

LG-ThinQ-Integration commented:
@joostlek 
I have a question too.
Do you mean we can't proceed more PRs, before the library is released?

joostlek commented:
That's the rule yes

LG-ThinQ-Integration commented:
@joostlek 
To make it clear, let me ask you one thing.
I think we are  doing just like you mentioned in 'switch.py'. Is there still any difference between your suggestion and our code?

joostlek commented:
Let me check

joostlek commented:
What exactly is a property in this case? Can you give an example?

joostlek commented:
Oh I think it's this
```json
    "operation": {
      "dehumidifierOperationMode": {
        "mode": ["r", "w"],
        "type": "enum",
        "value": {
          "r": ["POWER_OFF", "POWER_ON"],
          "w": ["POWER_OFF", "POWER_ON"]
        }
      }
    }
```

LG-ThinQ-Integration commented:
You can assume that a property matches a entity like a capability in smartthings integration.
But in some cases, two or more properties can be a single entity.

LG-ThinQ-Integration commented:
You understood correctly.

joostlek commented:
I think this is roughly what we want, except for the fact that we should remove the dynamic device creation (the `async_add_devices` function) since we don't have support for that yet. You did say that some entities are dependent on the model, where is this done? Looking at the code I would assume `create_properties`.

Apart from the fact that the entity_descriptions should move to `switch.py` (since the switch entity descriptions are not used anywhere else, so they should live in their own module) I think this is roughly what we need. I would suggest to decouple the function from the entity_descriptions and maybe just return a list of Properties, that can be linked to entity_descriptions in the platform files. I do the same in Withings https://github.com/home-assistant/core/blob/522a1e9d56ce8571b730c9ddf2108d58969b5044/homeassistant/components/withings/sensor.py#L70-L80

The only thing that I would question is the
```py
        property_info=PropertyInfo(
            key=propertyc.AIR_FAN_OPERATION_MODE,
            value_converter=value_to_power_state_converter,
        ),
```
part. I am not sure how it will eventually work with different data types (numbers/enum fields) but ideally we get the right value from the library already.

joostlek commented:
> But in some cases, two or more properties can be a single entity.

Can you give an example? If that is not in this PR yet, let's not dive too far into this, but it's nice to know what's up ahead down the road.

joostlek commented:
> You'll see them soon

I'll be on the lookout :)

joostlek commented:
And remove the things needed for dynamic device creation.

I think after that we can go back and forth with in depth review rounds and get everything ready :)

joostlek commented:
Awesome, if you have questions or doubts, feel free to ask me.

LG-ThinQ-Integration commented:
@joostlek 
I just want to make sure that even though you can download source from PyPI, we can proceed further after releasing the code on github. Is that right?

joostlek commented:
Nope, the source code needs to be available on GitHub (or any other git service like gitlab, bitbucket, whatever)

LG-ThinQ-Integration commented:
Understood.
The library code will be released very soon.

JangwonLeeLG committed:
Refactoring for reviews
1. Use coordinator extended class instead of LGDevice
2. Rename entity_helper.py to entity.py
3. Move entity description to each entity file
4. Remove dynamic device creation code

LG-ThinQ-Integration commented:
We updated belows
1. Use coordinator extended class instead of LGDevice
2. Rename entity_helper.py to entity.py
3. Move entity description to each entity file
4. Remove dynamic device creation code

joostlek commented:
Awesome! I will take a look

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+
+async def async_setup_entry(hass: HomeAssistant, entry: ThinqConfigEntry) -> bool:
+    """Set up an entry."""
+    access_token = entry.data.get(CONF_ACCESS_TOKEN)
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
```
These are always stored

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+    """Set up an entry."""
+    access_token = entry.data.get(CONF_ACCESS_TOKEN)
+    client_id = entry.data.get(CONF_CONNECT_CLIENT_ID)
+    country_code = entry.data.get(CONF_COUNTRY, DEFAULT_COUNTRY)
```
This one as well i believe

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+    except ThinQAPIException as exc:
+        raise ConfigEntryError(exc.message) from exc
+
+    if not device_list or not isinstance(device_list, Collection):
+        return
```
Why can the device list not be a Collection? That sounds to me that the API is responding in a weird way. If that's the case, the library should raise instead

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+    try:
+        device_list = await thinq_api.async_get_device_list()
+    except ThinQAPIException as exc:
+        raise ConfigEntryError(exc.message) from exc
```
We'd rather raise ConfigEntryNotReady because then it will retry

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+        hass.async_create_task(async_setup_device_coordinator(hass, thinq_api, device))
+        for device in device_list
+    ]
+    if task_list:
```
We already have a guard for having no devices in the list, so at this point we always have at least one, so this check is always true

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/__init__.py:
```diff
@@ -0,0 +1,129 @@
+            if coordinators:
+                coordinator_list += coordinators
+
+    # Register devices.
+    async_register_devices(hass, entry, coordinator_list)
+
+
+@callback
+def async_register_devices(
+    hass: HomeAssistant,
+    entry: ThinqConfigEntry,
+    coordinator_list: list[DeviceDataUpdateCoordinator],
+) -> None:
+    """Register devices to the device registry."""
+    device_registry = dr.async_get(hass)
+
+    for coordinator in coordinator_list:
+        device_entry = device_registry.async_get_or_create(
+            config_entry_id=entry.entry_id,
+            **coordinator.device_info,
+        )
+        _LOGGER.debug(
+            "Register device: device_id=%s, device_entry_id=%s",
+            coordinator.device_api.device_id,
+            device_entry.id,
+        )
+        entry.runtime_data[device_entry.id] = coordinator
```
When adding entities with a `device_info`, the device is automatically created. So we only need to do manual device creation when we either don't have the full info in the entity, or the device doesn't have entities (for example a hub). I don't think either are the case, so should we remove this?

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+    THINQ_PAT_URL,
+)
+
+SUPPORTED_COUNTRIES = [
+    SelectOptionDict(value=x.alpha_2, label=x.name) for x in pycountry.countries
+]
```
I'm wondering what countries we support. Do we support everything? If so, we do have a Country selector, which I think works the same like you have here, but with the added benefit that it will translate the country name (if I'm correct)

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+                client_id=connect_client_id,
+            ).async_get_device_list()
+        except ThinQAPIException as exc:
+            return self.async_abort(reason=exc.message)
```
We should not abort, we should return errors to the user. I will propose something in the next comment

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+            },
+        )
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow initiated by the user."""
+        # Get the PAT(Personal Access Token) and validate it.
+        if user_input is None:
```
Inline the validation function and inverse the flow here to make nice use of the errors

```py
errors: dict[str, str] = {}
if user_input is not None:
    ...validate input
    ...if catching error: `errors["base"] = "..."`
return self.async_show_form(....., errors=errors)
```

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,116 @@
+        country_code = str(user_input.get(CONF_COUNTRY, DEFAULT_COUNTRY))
+
+        # Check if PAT is already configured.
+        await self.async_set_unique_id(access_token)
```
Do we have something more unique for the user?

LG-ThinQ-Integration replied:
No we don't. 
If you have any better idea, please let me know.

joostlek replied:
Is there an endpoint on your side which we can use for fetching user data? How unique is this key?

LG-ThinQ-Integration replied:
Yes there is an end point. And we fetch data with PAT which is generated with a user account.
So we can guarantee PAT is unique to  identify a user.

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+            _LOGGER,
+            name=f"{DOMAIN}_{device_api.device_id}",
+        )
+        self._hass = hass
```
Hass is already stored in `self.hass`

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+        self._thinq_api = thinq_api
+        self._is_connected = True
+
+        # If sub_id is NONE_KEY("_") then it should be None.
+        self._sub_id = None if sub_id == NONE_KEY else sub_id
```
When can sub id be `_`?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+        # Get the api instance.
+        self._device_api = device_api.get_sub_device(self._sub_id) or device_api
+
+    @property
+    def thinq_api(self) -> ThinQApi:
+        """Returns the thinq api."""
+        return self._thinq_api
```
Might as well just store it in `self.thinq_api` from the beginning

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+        """Returns the thinq api."""
+        return self._thinq_api
+
+    @property
+    def device_api(self) -> ConnectBaseDevice:
+        """Returns the device api."""
+        return self._device_api
+
+    @property
+    def device_name(self) -> str:
+        """Returns the device name."""
+        return self._device_name
+
+    @property
+    def sub_id(self) -> str | None:
+        """Returns the device sub id."""
+        return self._sub_id
+
+    @property
+    def unique_id(self) -> str:
+        """Returns the unique id."""
+        return self._unique_id
```
Same for this, there's no real benefit of using private variables to then disclose them later in the public name

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+        """Check whether the device is connected or not."""
+        return self._is_connected
+
+    @is_connected.setter
+    def is_connected(self, connected: bool) -> None:
+        self._is_connected = connected
```
Why is this a setter?

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+            )
+        except ThinQAPIException as exc:
+            if exc.code == ThinQAPIErrorCodes.NOT_CONNECTED_DEVICE:
+                self.is_connected = False
+            return {}
+
+        # Full response into the device api.
+        self.device_api.set_status(data)
+        self.is_connected = True
+        return data
```
So CoordinatorEntity's are built to work together with the Coordinator. If the Coordinator raises `UpdateFailed` in `async_update_data`, it will automatically render all entities that are attached as `unavailable`, which is what we want.

If I understand the current code I think `is_connected` will be used for availability of the entities, but I think we should also be unavailable in the case when we can't successfully receive data from the device, for example when the cloud service is down or when someone's internet has problems

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+        self.is_connected = True
+        return data
+
+    def __str__(self) -> str:
+        """Return a string expression."""
+        return (
+            f"Coordinator:{self.device_name}"
+            f"(type={self.device_api.device_type}, id={self.device_api.device_id})"
+        )
```
No real use case for this imo

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+
+    # Create a list of sub-devices from the profile.
+    # Note that some devices may have more than two device profiles.
+    # In this case we should create multiple lg device instance.
+    # e.g. 'WashTower-Single-Unit' = 'WashTower{dryer}' + 'WashTower{washer}'.
```
Is this what you meant with the double device thingy?

LG-ThinQ-Integration replied:
Washtower is a single unit. But it has two devices(dryer, washer)
thinqconnect library send us as one device unit data for two devices at the same time in dictionary.
For user's own sake, we need to split them into 2 devices with ease to use.
If we don't split them, there might be many entities like power1, power2, mode1, mode2....
We think it could be a burden for a user when they control entities.

joostlek replied:
Ideally every physical device is its own device within Home Assistant. But I can also imagine that if you have a lot of appliances that it will be annoying with regards to entities and entity naming.

LG-ThinQ-Integration replied:
Let me explain how the washtower works.
When we get the washtower's properties, we get one dictionary which has two dictionary inside for dryer, washer (which has a different sub_id and a same device_id)
If one of dryer states changed, we send it with one of the coordinators with the proper sub_id (maybe first coordinator) and one of washer states changed, we send it with the other coordinator.
When it comes to creating a devices' coordinator(s),
if sub_ids exists create a coordinator for each sub_id and if there is no sub_id, create a single coordinator.

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,225 @@
+    )
+
+    # Create new device coordinator instances.
+    coordinator_list: list[DeviceDataUpdateCoordinator] = []
```
Can you elaborate on when this happens and what this means?

LG-ThinQ-Integration replied:
Because a washtower has two devices(dryer, washer), it needs two coordinators.
Because we assume a device is with one coordinator.

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+    # The property key for use in SDK must be snake_case string.
+    key: str
+
+    # Optional, if the value should be converted before calling api.
+    value_converter: Callable[[Any], Any] | None = None
```
Can you elaborate on what kind of data we get from the API? Is it possible to move the conversion to the library?

LG-ThinQ-Integration replied:
thinqconnect library gives us 'POWER_ON' string. but we need to convert it into true/false for Home Assistant.
Anyway, we are planning to move it into the library in the next modification.

joostlek replied:
Depends on how easy it is to move to the library. If the library can return multiple data types from that function I'd go with that. Otherwise just move the conversion function to the place where you set the state in the entity class, because it's the same every time

LG-ThinQ-Integration replied:
Understood. Let me move the conversion function to the place where I set the state.

joostlek replied:
If we move this, do we still need this PropertyInfo and extended entity description?

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+        self._property = property
+        self._attr_device_info = coordinator.device_info
+
+        # Set the unique key. If there exist a location, add the prefix location name.
+        unique_key = (
+            f"{entity_description.key}"
+            if property.location is None
+            else f"{property.location}_{entity_description.key}"
+        )
```
What's a location?

LG-ThinQ-Integration replied:
In case of a refrigerator, 'location' is one of 'freezer room' and 'refrigerator room'
'Target temperature' for those two rooms has same key.
That's why we create unique id with 'location'.

joostlek replied:
check

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+        )
+        self._attr_unique_id = f"{coordinator.unique_id}_{unique_key}"
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.coordinator.is_connected
```
Like said before, I don't think we should overwrite this

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+        """Return True if entity is available."""
+        return self.coordinator.is_connected
+
+    def get_property(self) -> ThinQProperty | None:
+        """Return the property corresponding to the feature."""
+        return self._property
```
Property is never None. Might as well just rename `_property` to `property`

LG-ThinQ-Integration replied:
Removed

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+            return
+
+        try:
+            await prop.async_post_value(value)
```
Can you tell me what kind of data we are expected to post?

LG-ThinQ-Integration replied:
example data> 'POWER_ON' string.

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,119 @@
+        self._update_status()
+        self.async_write_ha_state()
+
+    async def async_added_to_hass(self) -> None:
+        """Call when entity is added to hass."""
+        await super().async_added_to_hass()
+        self._handle_coordinator_update()
```
Instead I would recommend calling the `_update_status` at the end of the constructor

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,197 @@
+OPERATION_SWITCH_DESC: dict[ThinQProperty, ThinQSwitchEntityDescription] = {
+    ThinQProperty.AIR_FAN_OPERATION_MODE: ThinQSwitchEntityDescription(
+        key=ThinQProperty.AIR_FAN_OPERATION_MODE,
+        icon="mdi:power",
```
Please use icon translations for this

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,197 @@
+    ThinQProperty.AIR_FAN_OPERATION_MODE: ThinQSwitchEntityDescription(
+        key=ThinQProperty.AIR_FAN_OPERATION_MODE,
+        icon="mdi:power",
+        name="Power",
```
Name can be removed since you have translation keys.

If during testing you don't see the translated name, you probably forgot to generate translations. I don't have the command at me right now, let me know if you need it so you don't have to search yourself

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,197 @@
+    ),
+}
+
+AIR_PURIFIER_FAN_SWITCH: tuple[ThinQSwitchEntityDescription, ...] = (
+    OPERATION_SWITCH_DESC[ThinQProperty.AIR_FAN_OPERATION_MODE],
+)
+AIRPURIFIER_SWITCH: tuple[ThinQSwitchEntityDescription, ...] = (
+    OPERATION_SWITCH_DESC[ThinQProperty.AIR_PURIFIER_OPERATION_MODE],
+)
+DEHUMIDIFIER_SWITCH: tuple[ThinQSwitchEntityDescription, ...] = (
+    OPERATION_SWITCH_DESC[ThinQProperty.DEHUMIDIFIER_OPERATION_MODE],
+)
+HUMIDIFIER_SWITCH: tuple[ThinQSwitchEntityDescription, ...] = (
+    OPERATION_SWITCH_DESC[ThinQProperty.HUMIDIFIER_OPERATION_MODE],
+)
+SYSTEM_BOILER_SWITCH: tuple[ThinQSwitchEntityDescription, ...] = (
+    OPERATION_SWITCH_DESC[ThinQProperty.BOILER_OPERATION_MODE],
+)
```
Will there be more switches for these devices in the future?

It feels a bit repeating to first have a dict that can transform property into entity description, and then a list for switch per device type and then a way to get from device type to list?

LG-ThinQ-Integration replied:
Refactored

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,197 @@
+    def add_entities(
+        coordinators: Collection[DeviceDataUpdateCoordinator],
+    ) -> None:
+        """Add switch entities."""
+        for coordinator in coordinators:
+            descriptions = DEVIE_TYPE_SWITCH_MAP.get(coordinator.device_api.device_type)
+            if not isinstance(descriptions, tuple):
+                continue
+
+            entities: list[ThinQSwitchEntity] = []
+            for description in descriptions:
+                properties = create_properties(
+                    device_api=coordinator.device_api,
+                    key=description.key,
+                    children_keys=None,
+                    rw_type=PROPERTY_WRITABLE,
+                )
+                if not properties:
+                    continue
+
+                entities += [
+                    ThinQSwitchEntity(coordinator, description, prop)
+                    for prop in properties
+                ]
+
+            if entities:
+                for entity in entities:
+                    _LOGGER.debug(
+                        "[%s] Add %s:%s entity",
+                        coordinator.device_name,
+                        Platform.SWITCH,
+                        entity.entity_description.key,
+                    )
+                async_add_entities(entities)
+
+    add_entities(entry.runtime_data.values())
+
+    entry.async_on_unload(
+        async_dispatcher_connect(hass, THINQ_DEVICE_ADDED, add_entities)
+    )
```
Let's remove the automatic device adding for now

LG-ThinQ-Integration replied:
Do you mean below code?
---------------------------------------------
    entry.async_on_unload(
        async_dispatcher_connect(hass, THINQ_DEVICE_ADDED, add_entities)
    )
----------------------------------------------
If so,  it is removed.

joostlek replied:
And the callback

LG-ThinQ-Integration replied:
We already removed it.

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,197 @@
+
+        self._attr_is_on = self.get_value_as_bool()
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn on the switch."""
+        self._attr_is_on = True
+        self.async_write_ha_state()
+
+        _LOGGER.debug("[%s] async_turn_on", self.name)
+        await self.async_post_value(True)
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn off the switch."""
+        self._attr_is_on = False
+        self.async_write_ha_state()
+
+        _LOGGER.debug("[%s] async_turn_off", self.name)
+        await self.async_post_value(False)
```
Have you considered doing the post_value first and then asking the coordinator for a refresh so we sync the state again?

LG-ThinQ-Integration replied:
Modified

joostlek reviewed (changes requested):
I'm liking what I'm seeing! Put some more comments and questions.

The way I prefer to work with comments, is that if you think you fixed it, feel free to click resolve. But if it's a question I asked or if it's an ongoing discussion, please don't click resolve, because it's sometimes hard to find the comment back.

For bigger discussions, let's move the conversation to the PR comments so it's more visible.

JangwonLeeLG committed:
Refactoring for reviews

LG-ThinQ-Integration commented:
@joostlek 
I fixed some of reviews and commented for the rest of them.
Please check those again.

I have two things to talk to you.

1)
I have a favor to ask.
As you already know, every time we modify the library, we have to upgrade the library version.
We'd like to gather modifications and change the library all at once to reduce the count of upgrading version.
It would be very helpful if you give us the library-related reviews together.

2)
And I'd like you to ask about the progress.
As you saw in the email between Nabucasa and LGE, 
we expected to finish PRs in September and release 'thinq' integration on October 2nd.
And I heard we have to finish PRs at least 1 week before the release day.
As I know, we are not allowed to make a new PR before the parent PR is merged.
We have 10 entities in total.
Currently we are in the first PR and 9 PRs are left.
Do you think we are making proper progresses for the goal?

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,102 @@
+
+    def _get_default_country_code(self) -> str:
+        """Get the default country code based on config."""
+        return next(
+            (x for x in SUPPORTED_COUNTRIES if x == self.hass.config.country),
+            DEFAULT_COUNTRY,
+        )
```
Can't we do a `if self.hass.config.country not in SUPPORTED_COUNTRIES: return self.hass.config.country else: return DEFAULT_COUNTRY`?

LG-ThinQ-Integration replied:
Modified

joostlek commented on homeassistant/components/lgthinq/config_flow.py:
```diff
@@ -0,0 +1,102 @@
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            access_token = str(user_input[CONF_ACCESS_TOKEN])
+            country_code = str(user_input[CONF_COUNTRY])
```
No need to overwrite the type right?

LG-ThinQ-Integration replied:
Removed.

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,164 @@
+        # Get the api instance.
+        self.device_api = device_api.get_sub_device(self.sub_id) or device_api
+
+    @property
+    def device_info(self) -> dr.DeviceInfo:
+        """Return the device information."""
+        return dr.DeviceInfo(
+            identifiers={(DOMAIN, self.unique_id)},
+            manufacturer=COMPANY,
+            model=self.device_api.model_name,
+            name=self.device_name,
+        )
```
I think we can move this to `entity.py`, since we only use this there. You can use `_attr_device_info` to use in the constructor

LG-ThinQ-Integration replied:
Modified.

joostlek commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,164 @@
+    # e.g. 'WashTower-Single-Unit' = 'WashTower{dryer}' + 'WashTower{washer}'.
+    device_sub_ids = (
+        list(profile.keys())
+        if device_type == DeviceType.WASHTOWER and "property" not in profile
```
Only for the washtower?

LG-ThinQ-Integration replied:
Currently, yes.

joostlek commented on homeassistant/components/lgthinq/entity.py:
```diff
@@ -0,0 +1,100 @@
+        # Update initial status.
+        self._update_status()
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self.coordinator.last_update_success
```
```suggestion
```
Is the default

LG-ThinQ-Integration replied:
Removed.

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,159 @@
+            if not properties:
+                continue
+
+            entities += [
+                ThinQSwitchEntity(coordinator, description, prop) for prop in properties
+            ]
```
```suggestion
            entities.extend(
                ThinQSwitchEntity(coordinator, description, prop) for prop in properties
            )
```

LG-ThinQ-Integration replied:
Modified.

joostlek commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,159 @@
+            ]
+
+        if entities:
+            for entity in entities:
+                _LOGGER.debug(
+                    "[%s] Add %s:%s entity",
+                    coordinator.device_name,
+                    Platform.SWITCH,
+                    entity.entity_description.key,
+                )
```
This is already logged as far as I am aware

LG-ThinQ-Integration replied:
Removed.

joostlek commented on requirements_test_all.txt:
```diff
@@ -1441,6 +1441,9 @@ pycomfoconnect==0.5.1
 # homeassistant.components.coolmaster
 pycoolmasternet-async==0.2.2
 
+# homeassistant.components.lgthinq
+pycountry==23.12.11
+
```
Please run hassfest again

LG-ThinQ-Integration replied:
Updated

joostlek commented on tests/components/lgthinq/const.py:
```diff
@@ -0,0 +1,38 @@
+from typing import Final
+
+# A list of device type strings.
+AIR_CONDITIONER: Final[str] = "DEVICE_AIR_CONDITIONER"
+AIR_PURIFIER_FAN: Final[str] = "DEVICE_AIR_PURIFIER_FAN"
+AIR_PURIFIER: Final[str] = "DEVICE_AIR_PURIFIER"
+CEILING_FAN: Final[str] = "DEVICE_CEILING_FAN"
+COOKTOP: Final[str] = "DEVICE_COOKTOP"
+DEHUMIDIFIER: Final[str] = "DEVICE_DEHUMIDIFIER"
+DISH_WASHER: Final[str] = "DEVICE_DISH_WASHER"
+DRYER: Final[str] = "DEVICE_DRYER"
+HOOD: Final[str] = "DEVICE_HOOD"
+HOME_BREW: Final[str] = "DEVICE_HOME_BREW"
+HUMIDIFIER: Final[str] = "DEVICE_HUMIDIFIER"
+KIMCHI_REFRIGERATOR: Final[str] = "DEVICE_KIMCHI_REFRIGERATOR"
+MICROWAVE_OVEN: Final[str] = "DEVICE_MICROWAVE_OVEN"
+OVEN: Final[str] = "DEVICE_OVEN"
+PLANT_CULTIVATOR: Final[str] = "DEVICE_PLANT_CULTIVATOR"
+REFRIGERATOR: Final[str] = "DEVICE_REFRIGERATOR"
+ROBOT_CLEANER: Final[str] = "DEVICE_ROBOT_CLEANER"
+SYSTEM_BOILER: Final[str] = "DEVICE_SYSTEM_BOILER"
```
so why do we have this?

LG-ThinQ-Integration replied:
Removed

joostlek commented on tests/components/lgthinq/conftest.py:
```diff
@@ -0,0 +1,73 @@
+    ]
+
+
+@pytest.fixture(name="config_entry")
+def config_entry_fixture(
```
```suggestion
@pytest.fixture
def mock_config_entry(
```
We usually call it like this

LG-ThinQ-Integration replied:
We referred to other integrations including smartthings.
Those integrations are using it that way. 
Do we need to modify it anyway?

LG-ThinQ-Integration replied:
Modified

joostlek commented on tests/components/lgthinq/conftest.py:
```diff
@@ -0,0 +1,73 @@
+from tests.common import MockConfigEntry
+
+
+@pytest.fixture(name="country_code")
+def country_code_ficture() -> str:
+    """Return a mock country code."""
+    return MOCK_COUNTRY
+
+
+@pytest.fixture(name="connect_client_id")
+def connect_client_id_fixture() -> str:
+    """Return a mock connect client id."""
+    return f"{CLIENT_PREFIX}-{uuid.uuid4()!s}"
+
+
+@pytest.fixture(name="access_token")
+def access_token_fixture() -> str:
+    """Return a mock connect client id."""
+    return MOCK_PAT
```
You can just use name the fixtures the name so you don't have to rename them again

LG-ThinQ-Integration replied:
Many other integrations (smartthings, matter, google, media_player, etc) are using fixture renaming.
Do we need to remove it anyway?

LG-ThinQ-Integration replied:
Modified

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+async def test_config_flow(hass: HomeAssistant) -> None:
+    """Test that an thinq entry is normally created with valid values."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
```
```suggestion
        DOMAIN, context={"source": SOURCE_USER}
```
source is not a configuration key here

LG-ThinQ-Integration replied:
Modified

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+    await hass.async_block_till_done()
```
```suggestion
```

joostlek replied:
Please remove these after any config flow step we do, as they are most likely not needed

LG-ThinQ-Integration replied:
Removed

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        DOMAIN, context={CONF_SOURCE: SOURCE_USER}
+    )
+    await hass.async_block_till_done()
+    assert result["type"] == FlowResultType.FORM
```
```suggestion
    assert result["type"] is FlowResultType.FORM
```

joostlek replied:
And do this for every equality check with the FlowResultType enum

LG-ThinQ-Integration replied:
Modified.

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    )
+    await hass.async_block_till_done()
+    assert result["type"] == FlowResultType.FORM
+    assert result["step_id"] == SOURCE_USER
```
```suggestion
    assert result["step_id"] == "user"
```
Source isn't necisarily step_id

LG-ThinQ-Integration replied:
Modified

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+    assert result["type"] is FlowResultType.FORM
+    assert result["step_id"] == SOURCE_USER
+
+    with patch("homeassistant.components.lgthinq.config_flow.ThinQApi") as mock:
```
Can we create a central patch? Please check `airgradient` for an example. This eliminates any functions improperly patched

LG-ThinQ-Integration replied:
Refactored

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+        await hass.async_block_till_done()
+        assert result["type"] is FlowResultType.CREATE_ENTRY
+        assert result["title"] == MOCK_ENTRY_NAME
+        assert result["data"][CONF_ACCESS_TOKEN] == MOCK_PAT
+        assert result["data"][CONF_COUNTRY] == MOCK_COUNTRY
+
+        # Since a client_id is generated randomly each time,
+        # We can just check that the prefix format is correct.
+        client_id: str = result["data"][CONF_CONNECT_CLIENT_ID]
+        assert client_id.startswith(CLIENT_PREFIX)
```
please assert the full data dict. In order to test the client_id, just patch `uuid.uuid4` and return something static

LG-ThinQ-Integration replied:
Modified

joostlek commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,95 @@
+            data={CONF_COUNTRY: MOCK_COUNTRY},
+        )
+        await hass.async_block_till_done()
+        assert result["type"] is FlowResultType.ABORT
```
this should raise an error instead and the test should be repatched to work, and the test should end in CREATE_ENTRY so we test the flow is able to configure

LG-ThinQ-Integration replied:
Modified

joostlek reviewed (changes requested):
Can we add the test for the switch back? With this big integration I think it will add a lot of value

joostlek commented:
> It would be very helpful if you give us the library-related reviews together

I am not actively looking at the library, it's more that I see how it will be used and will propose ideas and stuff. In its current state I think it looks nice. In the future I can't guarantee any changes it needs. But I will try to let them know as early and grouped as possible :)

 > As I know, we are not allowed to create a new PR before the parent PR is merged

I think after this PR we have established on how the data is fetched and handled. So then (in theory) the other platforms will be more of a breeze (the readonly entites like binary sensor/sensor would of course be easier than a mutable entity with a lot of features like a fan or climate). I think after this you can open 1-3 PRs for the other platforms.

JangwonLeeLG committed:
Merge branch 'dev' of https://github.com/LG-ThinQ-Integration/core into pr1_1

JangwonLeeLG committed:
Update requirements

JangwonLeeLG committed:
Fix for reviews

LG-ThinQ-Integration commented:
Answer for this : https://github.com/home-assistant/core/pull/123860#discussion_r1736506190

In switch, every value_converter is same.
But in other entities, value_converter could be different for the same entity.
We are planning to create PR for sensor after this PR, and sensor entity has lots of value type.
In some cases, thinqconnect gives us string value, but in HomeAssistant, we need to convert it integer, and viceversa.
Can we keep this here?

JangwonLeeLG committed:
Modify tests for reviews

JangwonLeeLG committed:
Update for reviews

LG-ThinQ-Integration commented:
@joostlek 
I updated everything for your reviews.

joostlek commented:
We do strongly recommend it though. And at the moment its kinda strange because we do have all kinds of test fixtures but we don't use them. If you have a way to add the test I'm also happy to update the test myself to make it a valid test case, so we at least test something and we're sure it won't mess stuff up when we add the other platforms

joostlek commented:
> Answer for this : https://github.com/home-assistant/core/pull/123860#discussion_r1736506190
> 
> In switch, every value_converter is same.
> But in other entities, value_converter could be different for the same entity.
> We are planning to create PR for sensor after this PR, and sensor entity has lots of value type.
> In some cases, thinqconnect gives us string value, but in HomeAssistant, we need to convert it to integer, and viceversa.
> To handle several cases with ease, can we keep this here?

If not every entity would need this then we should not add it I think

LG-ThinQ-Integration commented:
> We do strongly recommend it though. And at the moment its kinda strange because we do have all kinds of test fixtures but we don't use them. If you have a way to add the test I'm also happy to update the test myself to make it a valid test case, so we at least test something and we're sure it won't mess stuff up when we add the other platforms

We updated tests.
We followed all your reviews.

JangwonLeeLG committed:
Remove property info and description class

LG-ThinQ-Integration commented:
> > Answer for this : [#123860 (comment)](https://github.com/home-assistant/core/pull/123860#discussion_r1736506190)
> > In switch, every value_converter is same.
> > But in other entities, value_converter could be different for the same entity.
> > We are planning to create PR for sensor after this PR, and sensor entity has lots of value type.
> > In some cases, thinqconnect gives us string value, but in HomeAssistant, we need to convert it to integer, and viceversa.
> > To handle several cases with ease, can we keep this here?
> 
> If not every entity would need this then we should not add it I think

We removed PropertyInfo class and related description code.

joostlek commented:
Awesome! I will take another look. Before what time are you still online to open follow up PRs? To avoid waiting the weekend so you already have feedback on Monday to work on (or not if everything goes well)

LG-ThinQ-Integration commented:
Do you mean the next PRs?
It so, it's happy to hear that.
If this PR goes OK, we are planning to proceed PRs for some of sensor, climate and mqtt.
If possible, we will make PRs on Monday.

LG-ThinQ-Integration commented:
If there are reviews for minor changes, I can respond to you anytime even during weekend.
But if it is a big change, I think I could handle next week.
By the way, I'll be online for 2 hours from now.

joostlek commented:
Okay then let's not hurry now. The timezone difference is nice because you have the mornings to fix stuff and I have the mornings to review that again

joostlek commented:
Oh and I haven't taken a look at the new code yet. I was just wondering if that would be easier for you :)

LG-ThinQ-Integration commented:
> Oh and I haven't taken a look at the new code yet. I was just wondering if that would be easier for you :)

I'm ready to get your reviews anytime.  😊 
I'm grateful for your kindness.

LG-ThinQ-Integration committed:
Update tests/components/lgthinq/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update tests/components/lgthinq/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/entity.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update tests/components/lgthinq/test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

JangwonLeeLG committed:
Update for reviews

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration committed:
Update homeassistant/components/lgthinq/switch.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,120 @@
+        if not isinstance(descriptions, tuple):
+            continue
+
+        entities: list[ThinQSwitchEntity] = []
```
Modified

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,153 @@
+        try:
+            await coordinator.async_config_entry_first_refresh()
+        except ConfigEntryNotReady:
+            coordinator.data = {}
```
Modified.

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/coordinator.py:
```diff
@@ -0,0 +1,153 @@
+    hass: HomeAssistant, thinq_api: ThinQApi, device: dict[str, Any]
+) -> list[DeviceDataUpdateCoordinator] | None:
+    """Create DeviceDataUpdateCoordinator and device_api per device."""
+    device_id = device.get("deviceId")
+    if not device_id:
+        _LOGGER.error("Failed to setup device: no device id")
+        return None
+
+    device_info = device.get("deviceInfo")
+    if not device_info:
+        _LOGGER.error("Failed to setup device(%s): no device info", device_id)
+        return None
```
Unnecessary.
Removed.

LG-ThinQ-Integration replied:
Actually there is no error case.
But because type is Any, we check 'device_id' and 'device_info' to escape ruff check issue.
If you have any suggestion, please let me know.

joostlek replied:
`device["deviceId"]`

In case there are more places where we do extra stuff to make mypy happy, feel free to ask help. I'd rather help with this than to have checks in the code for stuff that never happens

LG-ThinQ-Integration replied:
Thank you for your suggestion.
Modified.

JangwonLeeLG committed:
Update for reviews

JangwonLeeLG committed:
Fix ruff issues

JangwonLeeLG committed:
Fix ruff check

LG-ThinQ-Integration commented on tests/components/lgthinq/conftest.py:
```diff
@@ -0,0 +1,71 @@
+def mock_thinq_api() -> Generator[AsyncMock]:
+    """Mock a thinq api."""
+    with (
+        patch("thinqconnect.ThinQApi", autospec=True) as mock_api,
```
Moved

LG-ThinQ-Integration commented on tests/components/lgthinq/common.py:
```diff
@@ -0,0 +1,19 @@
+
+def mock_thinq_api_response(
+    *,
+    status: int = 400,
```
Changed to 200

LG-ThinQ-Integration commented on homeassistant/components/lgthinq/switch.py:
```diff
@@ -0,0 +1,120 @@
+    for coordinator in entry.runtime_data.values():
+        descriptions = DEVIE_TYPE_SWITCH_MAP.get(coordinator.device_api.device_type)
+        if not isinstance(descriptions, tuple):
+            continue
```
Modified.

JangwonLeeLG committed:
Fix for reviews

JangwonLeeLG committed:
Fix ruff check

LG-ThinQ-Integration commented on tests/components/lgthinq/test_config_flow.py:
```diff
@@ -0,0 +1,66 @@
+    )
+    await hass.async_block_till_done()
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] is not None
```
Modified

LG-ThinQ-Integration committed:
Fix for reviews

LG-ThinQ-Integration commented:
@joostlek 
Thank you for you reviews again.
I updated everything you mentioned.

LG-ThinQ-Integration commented:
Prettier? Please let me know to fix failures.

joostlek commented:
Hmm, can't seem to see how to run them manually. Maybe it helps if you have pre-commit installed correctly and then just add some new lines to the end of the file and then commit. It will then pick it up. Also make sure to run the hassfest checks

LG-ThinQ-Integration committed:
Merge branch 'dev' into pr1_1

LG-ThinQ-Integration commented:
> Hmm, can't seem to see how to run them manually. Maybe it helps if you have pre-commit installed correctly and then just add some new lines to the end of the file and then commit. It will then pick it up. Also make sure to run the hassfest checks

I found it.
- pre-commit run --hook-stage manual prettier --all-files

And let me run hassfest checks also.

LG-ThinQ-Integration committed:
Fix prettier failure and hassfest failure

LG-ThinQ-Integration commented:
I fixed prettier issues and ran hassfest also.

LG-ThinQ-Integration commented:
@joostlek 
All passed.
Can I open  the PR for review?

LG-ThinQ-Integration commented:
Hmm 2 failures after success.
Codecove?
How can I fix it?

joostlek commented:
Add tests :P

But we can do that in a follow up :)

LG-ThinQ-Integration commented:
Thank you~!!
